[WIP] AI Adjustments

Textor44 opened the pull request:
:cl: Textor
rscadd: With the controversy about privacy and artificial intelligence, all security cameras now blink when an AI is actively looking through them.
rscadd: You can now block cameras with sticky notes and duct tape.
rscadd: Intercoms now blink and indicate when their mics are set to permanent broadcast.
rscadd: Traitors now have access to the communications blackout event to disable comms. This can still be mitigated by resetting the servers in telecomms.
rscadd: Cameras now have their own power channel on an APC. They also now have battery backups so they don't immediately shut off when the power is cut. The cameras use more power when the AI is staring through them, so unpowered areas will lose camera coverage faster when the AI is staring at the area.
tweak: Synthetics bolting doors now takes 5 seconds. Electrifying doors now takes 10 seconds. Both give indicators to people nearby that these actions have been taken to allow for people to respond to what's happening.
/:cl:

This PR is an attempt to address some of the main complaints people have mechanically with the AI. There are more things planned, and I will be taking additional suggestions.

- [x] Revert removing tracking ability from the AI
- [x] Delay bolting and electrifying doors.
- [x] Fix the duct tape/post-it notes issue
- [x] Add cameras to APC, 
- [ ] Set cameras to be the last thing to shut off when power is out.
- [ ] Make cameras be affected by the APC instead of just being pretty parts of the APC's interface.
- [ ] Balance camera power usage.
- [ ] Add power reroute interface to allow the AI to transfer energy from any connected APCs to any other connected APCs to power areas.

Textor44 committed:
[WIP] Anti-AI Measure Additions

:cl: Textor
rscadd: WIth the controversy about privacy and artificial intelligence, all security cameras now blink when an AI is actively looking through them.
rscadd: Someone had the bright idea to cover camera lenses up with duct tape and sticky notes. The Expeditionary Corps would like to remind all personnel that they should not do this for security and safety reasons.
/:cl:

This PR is an attempt to address some of the main complaints people have mechanically with the AI. There are more things planned, and I will be taking additional suggestions.

Sunbeamstress commented:
I'm not a dev person, but can I request that you change the second changelog note to simply say you can cover cameras with those objects? I like the way you phrased it but it may not be obvious it's an actual feature addition.

PsiOmegaDelta commented:
The "Expeditionary Corps"-part is indeed superfluous for a changelog entry.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -123,6 +130,16 @@
+	queue_icon_update()
+
+/obj/machinery/camera/proc/ai_stop_watching()
+	ai_watching = 0
```
If multiple AIs are within range of the same camera this value will be incorrectly set as false.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -123,6 +130,16 @@
+/obj/machinery/camera/proc/set_ai_watching(var/mob/observer/eye/aiEye/ai)
+	if((stat & BROKEN) || (stat & EMPED))
+		return
+	ai_watching = 1
```
If an AI is destroyed, along with its eye, the camera status won't update correctly.

Textor44 replied:
This is now handled when the eye is destroyed.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -42,6 +47,35 @@
+			C.set_ai_watching(eye)
+			eye.add_camera(C)
+	return
+/proc/check_ai_in_range(var/mob/observer/eye/aiEye/eye, var/obj/machinery/camera/C)
```
This proc just adds unnecessary overhead in an already questionable loop. Just inline and replace it with `get_dist()` instead.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -42,6 +47,35 @@
+
+/proc/ai_near_camera(var/mob/observer/eye/aiEye/eye)
+	for(var/obj/machinery/camera/C in range(8, eye))
+		if(C.can_use())
```
Cameras are never added to AI eye lists if they can be used again while the AI is in range.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -42,6 +47,35 @@
 		return 1
 
+/mob/observer/eye/aiEye/proc/add_camera(var/obj/machinery/camera/C)
+	if(!camera_list.Find(C))
```
`Find()` is one of, if not the, most expensive proc in BYOND to verify if an instance is in a list.
Just do `camera_list |= C`

PsiOmegaDelta replied:
In fact, consider using the `LAZY`-macros instead.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -42,6 +47,35 @@
+	. = ..()
+
+/proc/ai_near_camera(var/mob/observer/eye/aiEye/eye)
+	for(var/obj/machinery/camera/C in range(8, eye))
```
Checking for cameras in range every time a very speedy and mobile object moves? I'm having performance concerns.

PsiOmegaDelta commented:
Also consider adding more newlines between procs. It was a bother to read the diff in `eye.dm`

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -166,7 +183,12 @@
+	if(blocked)
+		blocked = 0
+		to_chat(user, "<span class='notice'>You remove something obstructing [src]'s lens.</span>")
+		set_broken(FALSE)
```
This will magically repair cameras with bad wiring.

Textor44 committed:
Makes changes, fixes icon state name for blocked cameras

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -15,13 +15,18 @@
 	name = "Inactive AI Eye"
 	name_sufix = "AI Eye"
 	icon_state = "AI-eye"
+	var/list/camera_list = list()
```
This list prevents a clean GC of deleted cameras.
You'll need to register to the deletion event and remove cameras from the list as necessary.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/ai/eye.dm:
```diff
@@ -84,13 +85,14 @@
 	var/obj/machinery/hologram/holopad/holo = null
 
 /mob/living/silicon/ai/proc/destroy_eyeobj(var/atom/new_eye)
+	eyeobj.purge_cameras()
```
This should be handled in the eye's `/Destroy()`, not here.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -166,7 +183,12 @@
+	if(blocked)
+		blocked = 0
+		to_chat(user, "<span class='notice'>You remove something obstructing [src]'s lens.</span>")
+		set_status(1, 1)
```
I find no version of `set_status()` that takes a second argument.

Roland410 commented:
I'm fine with the camera blinking, but the duct tape part makes hacking cameras obsolete.

Textor44 commented:
> I'm fine with the camera blinking, but the duct tape part makes hacking cameras obsolete.

It'd be easier to fix (literally walk up to the camera and rip the tape off) as well. If you want to make it so some random person couldn't come by and fix what you did, you'd have to hack it properly or smash it.

I'm trying to provide solutions to the problems people seem to have with the AI, and a lot of the debate has been all about cameras and disabling them/knowing when they are being monitored.

More is coming that isn't strictly camera stuff, too.

WhiteHusky commented on code/game/machinery/camera/camera.dm:
```diff
@@ -211,17 +231,26 @@
-			else to_chat(O, "<b><a href='byond://?src=\ref[O];track2=\ref[O];track=\ref[U];trackname=[U.name]'>[U]</a></b> holds \a [itemname] up to one of your cameras ...")
-			O << browse(text("<HTML><HEAD><TITLE>[]</TITLE></HEAD><BODY><TT>[]</TT></BODY></HTML>", itemname, info), text("window=[]", itemname))
+	else if (can_use() && isliving(user) && !blocked)
+		if(istype(W, /obj/item/weapon/paper) || istype(W, /obj/item/weapon/ducttape) || istype(W, /obj/item/weapon/tape_roll))
```
`istype(W, /obj/item/weapon/paper)`
This looks like it'll prevent people from showing the AI a piece of paper. Might want to check intent so if someone is `I_HELP` they show the paper while everything else will obscure the camera.

Textor44 replied:
that was an oversight, it should be paper_sticky, not paper.

WhiteHusky commented on code/game/machinery/camera/camera.dm:
```diff
@@ -211,17 +231,26 @@
+			set_status(0, 1)
+			to_chat(user, "<span class='notice'>You stick [W] on [src]'s lens, causing an obstruction.</span>")
+			if(!istype(W, /obj/item/weapon/tape_roll))
+				qdel(W)
```
Consider having the ~~paper~~ item moved into the camera so when the obstruction is removed, it'll be taken out from the camera. Leaves something for the detectives to investigate or someone to retrieve a document they misplaced on the camera.

Textor44 committed:
Makes more requested changes, makes intercoms blink when they are actively transmitting, adds communications blackout device for traitors

Known bugs:
-Tape/post-its aren't updating icons correctly
-Disabled cameras aren't removing their visual range from the AI, pretty sure it isn't related to my changes, but it is still being investigated.

Roland410 commented:
Hacking cams is easy, and fixing them as well. The purpose of disabling cameras is to not give the AI/security vision into the room, in case of an AI, you can do whatever until said camera gets fixed. Duct tape and sticky notes are easy to find and use in high numbers, effectively being able to block entire hallways or decks.
Also probably should wait with merging until Cakey's PR gets closed/merged.

Textor44 commented:
> Hacking cams is easy, and fixing them as well. The purpose of disabling cameras is to not give the AI/security vision into the room, in case of an AI, you can do whatever until said camera gets fixed. Duct tape and sticky notes are easy to find and use in high numbers, effectively being able to block entire hallways or decks.

I'll probably add a timer to it to make it take a couple seconds to do each camera to slow things down a bit. And I agree, hacking/destroying cameras _is_ easy to do, but apparently a lot of antags find it way too hard.

> Also probably should wait with merging until Cakey's PR gets closed/merged.

If it gets merged this will get closed out. This is here to present an alternative to absolute removal.

Textor44 committed:
Adds closing span tag to message.

Roland410 commented:
I don't think we should nerf something because antags are lazy, but at the very least put a timer on it, sounds like an acceptable solution.

afterthought2 commented:
The diff is not for discussing general attitude towards changes. Further discussion here should be moved to discord or the forums; disregarding this message will result in repo warnings or bans.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -211,17 +229,30 @@
+			else
+				var/obj/item/weapon/ducttape/T = new /obj/item/weapon/ducttape(src)
+				T.add_fingerprint(user)
+				contents += T
```
Need to handle these added items in the camera's `/Destroy()`

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -237,31 +268,40 @@
 	if(choice != 1)
 		return
-
+	log_and_message_admins("blocking = [blocking]")
```
A `user` is supplied, make this `log_and_message_admins(message, user)` (otherwise it defaults to `usr`)
Also the final message is "[user.name] [message]" making the above message turn into "McUrist blocking = [0/1]"

Textor44 replied:
It's debug messaging, and will be removed.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -237,31 +268,40 @@
-		if(user)
+		if(user && blocking)
+			log_and_message_admins("Excuted user+blocking")
+			visible_message("<span class='notice'> [user] has put something in front of [src]'s lens.</span>")
```
Odd space between the `>` and `[user]`

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -237,31 +268,40 @@
+			log_and_message_admins("Excuted user+blocking")
+			visible_message("<span class='notice'> [user] has put something in front of [src]'s lens.</span>")
+			icon_state = "[initial(icon_state)]_blocked"
+			add_hiddenprint(user)			
```
Why hidden prints?

Textor44 replied:
that's what was already there for how the other cases were handled. I'll move it over to the normal fingerprint proc.

PsiOmegaDelta commented on code/game/machinery/camera/camera.dm:
```diff
@@ -237,31 +268,40 @@
+			icon_state = "[initial(icon_state)]_blocked"
+			add_hiddenprint(user)			
+		else if(user)
+			log_and_message_admins("Executed user")
```
Resulting message: "McUrist Executed user"

Textor44 replied:
more debug messaging, and will be removed.

PsiOmegaDelta commented on code/modules/mob/observer/freelook/eye.dm:
```diff
@@ -102,6 +102,9 @@
 			return
 		return loc
 
+/mob/observer/eye/proc/purge_cameras()
```
Unused proc.

Textor44 committed:
Removes tracking abilities, applies delays to shocking and bolting of doors by the AI

babydoll commented:
Create a thread in the gameplay changes subforum using the template in this thread: https://forums.baystation12.net/threads/gameplay-change-subforum-rules.5838/

Textor44 commented:
Discussion thread: https://forums.baystation12.net/threads/ai-adjustments-24299.7388/

WhiteHusky commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -898,13 +899,22 @@ About the new airlock wires panel:
 				to_chat(usr, "The door bolts have been raised.")
 		if("electrify_temporary")
+			if(isAI(usr) && do_after(usr, 80, src))
+				visible_message("<span class='warning'>Electricity appears to be building up in the [src]!</span>")
```
`SPAN_WARNING("\The [src] makes a concerning electrical hum!")`

WhiteHusky commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -525,6 +525,7 @@ About the new airlock wires panel:
 		message = text("The door is unpowered - Cannot electrify the door.")
 		src.electrified_until = 0
 	else if(!duration && electrified_until != 0)
+		visible_message("<span class='notice'>The [src]'s charge dissipates.</span>'")
```
`SPAN_NOTICE("The humming from \the [src] dies down.")`

WhiteHusky reviewed (changes requested):
I'm against directly hinting to what's happening, more of just describing the sound and having the player interpret what it means.

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into AI-Tweaks

babydoll commented:
I've posted my comments in the thread, and they'll need to be addressed before this can be merged. Textor isn't required to follow up, but anyone who picks up the baton if he doesn't want to will be similarly held.

Textor44 commented:
I'll start working on it more tomorrow, and try to make the adjustments requested.

BlueNexus commented on code/game/machinery/camera/tracking.dm:
```diff
@@ -152,37 +137,12 @@
+	if(U.eyeobj)
+		U.eyeobj.setLoc(get_turf(target), 0)
+	else
+		view_core()
```
I know this would be overengineering, but would 
```
    if(!(U?.eyeobj?.setLoc(get_turf(target),0)))
        view_core()
```
Work here? Just found that operator and I'm curious if that's how it works.

Textor44 replied:
probably, but it makes the code harder to understand.

Textor44 committed:
Makes requested changes, restores follow verb.

Textor44 committed:
Fixes duct tape/sticky notes, Lays groundwork for APC controlled cameras

-Duct tape and sticky notes work correctly now.
-Adds a new power network for cameras
-Adds power cells into cameras
-Makes cameras operate off of their power cells if power is offline or the circuit is shut off.
-Cameras use more power when the AI is looking through them
-Cameras use more power when the cells need recharging.
-APC overlays updated to add camera status lights
-APC interface updated to control camera power.

Textor44 committed:
Fixes span tag mismatch.

Textor44 committed:
Merge remote-tracking branch 'upstream/dev' into AI-Tweaks

SierraKomodo commented:
> tweak: Synthetics bolting doors now takes 5 seconds. Electrifying doors now takes 10 seconds. Both give indicators to people nearby that these actions have been taken to allow for people to respond to what's happening.

Will this also have any effect on bolting/unbolting of doors in airlocks or with multitools/wire cutting?

Textor44 commented:
> > tweak: Synthetics bolting doors now takes 5 seconds. Electrifying doors now takes 10 seconds. Both give indicators to people nearby that these actions have been taken to allow for people to respond to what's happening.
> 
> Will this also have any effect on bolting/unbolting of doors in airlocks or with multitools/wire cutting?

Synthetics don't have to wait for the unbolt command to work, and it would not affect anyone using tools on the door or airlock controllers.

babydoll commented:
I'm going to close this now since it's been over a month since I made my decision, and as far as I can tell the PR has not yet reflected the provisions I made. Let me or another developer know if you'd like this re-opened.
